File: src/card_variant.py
"""Level-based Pokemon TCG holo tiers matching pokemon-cards-css categories."""

from dataclasses import dataclass

STATIC = "/static/vendor/pokemon-cards-css/img"


@dataclass(frozen=True)
class Variant:
    """One holo tier — wire attributes match pokemon-cards-css Card.svelte."""

    name: str
    badge: str
    css_class: str
    data_rarity: str
    supertype: str
    subtypes: str
    trainer_gallery: bool = False
    masked: bool = False
    holo: bool = False
    sparkles: bool = False
    foil_style: str = ""
    mask_style: str = ""
    card_number: str = ""

    @property
    def rainbow_frame(self) -> bool:
        return self.data_rarity in ("rare rainbow", "rare rainbow alt")

    @property
    def gold_frame(self) -> bool:
        return self.data_rarity == "rare secret"


COMMON = Variant(
    name="Common",
    badge="",
    css_class="hpk-v-common",
    data_rarity="common",
    supertype="trainer",
    subtypes="supporter",
    holo=False,
)

REVERSE_HOLO = Variant(
    name="Reverse Holo",
    badge="reverse",
    css_class="hpk-v-reverse",
    data_rarity="uncommon reverse holo",
    supertype="trainer",
    subtypes="supporter",
    masked=True,
    holo=True,
    foil_style=f"url({STATIC}/angular.png)",
    mask_style=f"url({STATIC}/wave.png)",
)

HOLO_RARE = Variant(
    name="Holo Rare",
    badge="holo",
    css_class="hpk-v-holo",
    data_rarity="rare holo",
    supertype="trainer",
    subtypes="supporter",
    holo=True,
)

COSMOS = Variant(
    name="Cosmos Holo",
    badge="cosmos",
    css_class="hpk-v-cosmos",
    data_rarity="rare holo cosmos",
    supertype="trainer",
    subtypes="supporter",
    holo=True,
)

AMAZING = Variant(
    name="Amazing Rare",
    badge="amazing",
    css_class="hpk-v-amazing",
    data_rarity="amazing rare",
    supertype="trainer",
    subtypes="supporter",
    masked=True,
    holo=True,
    sparkles=True,
    foil_style=f"url({STATIC}/glitter.png)",
    mask_style=f"url({STATIC}/geometric.png)",
)

RADIANT = Variant(
    name="Radiant",
    badge="radiant",
    css_class="hpk-v-radiant",
    data_rarity="radiant rare",
    supertype="trainer",
    subtypes="radiant",
    holo=True,
    sparkles=True,
)

TRAINER_GALLERY = Variant(
    name="Trainer Gallery",
    badge="TG",
    css_class="hpk-v-tg",
    data_rarity="rare holo",
    supertype="trainer",
    subtypes="supporter",
    trainer_gallery=True,
    holo=True,
    card_number="TG01",
)

POKEMON_V = Variant(
    name="Pokemon V",
    badge="V",
    css_class="hpk-v-v",
    data_rarity="rare holo v",
    supertype="pokémon",
    subtypes="basic v",
    holo=True,
)

V_FULL_ART = Variant(
    name="V Full Art",
    badge="V FA",
    css_class="hpk-v-vfa",
    data_rarity="rare ultra",
    supertype="pokémon",
    subtypes="basic v",
    holo=True,
    sparkles=True,
    foil_style=f"url({STATIC}/illusion.png)",
)

V_ALT_ART = Variant(
    name="V Alt Art",
    badge="V AA",
    css_class="hpk-v-vaa",
    data_rarity="rare ultra",
    supertype="pokémon",
    subtypes="basic v",
    holo=True,
    sparkles=True,
    foil_style=f"url({STATIC}/illusion.png)",
    mask_style=f"url({STATIC}/illusion-mask.png)",
    masked=True,
)

VMAX = Variant(
    name="VMax",
    badge="VMAX",
    css_class="hpk-v-vmax",
    data_rarity="rare holo vmax",
    supertype="pokémon",
    subtypes="vmax",
    holo=True,
    sparkles=True,
)

VMAX_RAINBOW = Variant(
    name="VMax Rainbow",
    badge="VMAX ★",
    css_class="hpk-v-vmax-r",
    data_rarity="rare rainbow",
    supertype="pokémon",
    subtypes="vmax",
    holo=True,
    sparkles=True,
)

VSTAR = Variant(
    name="VStar",
    badge="VSTAR",
    css_class="hpk-v-vstar",
    data_rarity="rare holo vstar",
    supertype="pokémon",
    subtypes="vstar",
    holo=True,
    sparkles=True,
)

SECRET_GOLD = Variant(
    name="Secret Gold",
    badge="★ gold",
    css_class="hpk-v-gold",
    data_rarity="rare secret",
    supertype="trainer",
    subtypes="supporter",
    holo=True,
    sparkles=True,
    foil_style=f"url({STATIC}/geometric.png)",
)

VARIANTS: tuple[Variant, ...] = (
    COMMON,
    REVERSE_HOLO,
    HOLO_RARE,
    COSMOS,
    AMAZING,
    RADIANT,
    TRAINER_GALLERY,
    POKEMON_V,
    V_FULL_ART,
    V_ALT_ART,
    VMAX,
    VMAX_RAINBOW,
    VSTAR,
    SECRET_GOLD,
)

_TIER_THRESHOLDS: tuple[tuple[int, Variant], ...] = (
    (94, SECRET_GOLD),
    (87, VSTAR),
    (80, VMAX_RAINBOW),
    (73, VMAX),
    (66, V_ALT_ART),
    (59, V_FULL_ART),
    (52, POKEMON_V),
    (45, TRAINER_GALLERY),
    (38, RADIANT),
    (31, AMAZING),
    (24, COSMOS),
    (17, HOLO_RARE),
    (10, REVERSE_HOLO),
)


def variant_for_level(level: int) -> Variant:
    """Return the pokemon-cards-css tier for a Hugging Face profile level."""
    for threshold, variant in _TIER_THRESHOLDS:
        if level >= threshold:
            return variant
    return COMMON


def foil_mask_style(variant: Variant) -> str:
    """Inline --foil / --mask for card__front when the tier needs masking."""
    parts: list[str] = []
    if variant.foil_style:
        parts.append(f"--foil:{variant.foil_style}")
    if variant.mask_style:
        parts.append(f"--mask:{variant.mask_style}")
    return ";".join(parts)

File: web/src/lib/card-variant.ts
/** Face layout templates — each maps to a pre-made overlay + pokemon-cards-css wire attrs */

export type FaceTemplate =
  | "trainer"
  | "pokemon-v"
  | "full-art"
  | "vmax"
  | "rainbow"
  | "vstar"
  | "secret";

export type CardVariant = {
  name: string;
  badge: string;
  dataRarity: string;
  supertype: "trainer" | "pokémon";
  subtypes: string;
  trainerGallery: boolean;
  masked: boolean;
  faceTemplate: FaceTemplate;
  foil?: string;
  mask?: string;
};

const IMG = "/pokemon-cards-css/img";

const COMMON: CardVariant = {
  name: "Common",
  badge: "",
  dataRarity: "common",
  supertype: "trainer",
  subtypes: "supporter",
  trainerGallery: false,
  masked: false,
  faceTemplate: "trainer",
};

const REVERSE_HOLO: CardVariant = {
  name: "Reverse Holo",
  badge: "reverse",
  dataRarity: "uncommon reverse holo",
  supertype: "trainer",
  subtypes: "supporter",
  trainerGallery: false,
  masked: true,
  faceTemplate: "trainer",
  foil: `${IMG}/angular.png`,
  mask: `${IMG}/wave.png`,
};

const HOLO_RARE: CardVariant = {
  name: "Holo Rare",
  badge: "holo",
  dataRarity: "rare holo",
  supertype: "trainer",
  subtypes: "supporter",
  trainerGallery: false,
  masked: false,
  faceTemplate: "trainer",
};

const COSMOS: CardVariant = {
  name: "Cosmos Holo",
  badge: "cosmos",
  dataRarity: "rare holo cosmos",
  supertype: "trainer",
  subtypes: "supporter",
  trainerGallery: false,
  masked: false,
  faceTemplate: "trainer",
};

const AMAZING: CardVariant = {
  name: "Amazing Rare",
  badge: "amazing",
  dataRarity: "amazing rare",
  supertype: "trainer",
  subtypes: "supporter",
  trainerGallery: false,
  masked: true,
  faceTemplate: "trainer",
  foil: `${IMG}/glitter.png`,
  mask: `${IMG}/geometric.png`,
};

const RADIANT: CardVariant = {
  name: "Radiant",
  badge: "radiant",
  dataRarity: "radiant rare",
  supertype: "trainer",
  subtypes: "radiant",
  trainerGallery: false,
  masked: false,
  faceTemplate: "trainer",
};

const TRAINER_GALLERY: CardVariant = {
  name: "Trainer Gallery",
  badge: "TG",
  dataRarity: "rare holo",
  supertype: "trainer",
  subtypes: "supporter",
  trainerGallery: true,
  masked: false,
  faceTemplate: "trainer",
};

const POKEMON_V: CardVariant = {
  name: "Pokemon V",
  badge: "V",
  dataRarity: "rare holo v",
  supertype: "pokémon",
  subtypes: "basic v",
  trainerGallery: false,
  masked: false,
  faceTemplate: "pokemon-v",
};

const V_FULL_ART: CardVariant = {
  name: "V Full Art",
  badge: "V FA",
  dataRarity: "rare ultra",
  supertype: "pokémon",
  subtypes: "basic v",
  trainerGallery: false,
  masked: false,
  faceTemplate: "full-art",
  foil: `${IMG}/illusion.png`,
};

const V_ALT_ART: CardVariant = {
  name: "V Alt Art",
  badge: "V AA",
  dataRarity: "rare ultra",
  supertype: "pokémon",
  subtypes: "basic v",
  trainerGallery: false,
  masked: true,
  faceTemplate: "full-art",
  foil: `${IMG}/illusion.png`,
  mask: `${IMG}/illusion-mask.png`,
};

const VMAX: CardVariant = {
  name: "VMax",
  badge: "VMAX",
  dataRarity: "rare holo vmax",
  supertype: "pokémon",
  subtypes: "vmax",
  trainerGallery: false,
  masked: false,
  faceTemplate: "vmax",
};

const VMAX_RAINBOW: CardVariant = {
  name: "VMax Rainbow",
  badge: "VMAX ★",
  dataRarity: "rare rainbow",
  supertype: "pokémon",
  subtypes: "vmax",
  trainerGallery: false,
  masked: false,
  faceTemplate: "rainbow",
};

const VSTAR: CardVariant = {
  name: "VStar",
  badge: "VSTAR",
  dataRarity: "rare holo vstar",
  supertype: "pokémon",
  subtypes: "vstar",
  trainerGallery: false,
  masked: false,
  faceTemplate: "vstar",
};

const SECRET_GOLD: CardVariant = {
  name: "Secret Gold",
  badge: "★ gold",
  dataRarity: "rare secret",
  supertype: "trainer",
  subtypes: "supporter",
  trainerGallery: false,
  masked: false,
  faceTemplate: "secret",
  foil: `${IMG}/geometric.png`,
};

const THRESHOLDS: [number, CardVariant][] = [
  [94, SECRET_GOLD],
  [87, VSTAR],
  [80, VMAX_RAINBOW],
  [73, VMAX],
  [66, V_ALT_ART],
  [59, V_FULL_ART],
  [52, POKEMON_V],
  [45, TRAINER_GALLERY],
  [38, RADIANT],
  [31, AMAZING],
  [24, COSMOS],
  [17, HOLO_RARE],
  [10, REVERSE_HOLO],
];

export function variantForLevel(level: number): CardVariant {
  for (const [threshold, variant] of THRESHOLDS) {
    if (level >= threshold) return variant;
  }
  return COMMON;
}

/** Mini-card shell class for binder follower pockets */
export function miniCardClass(variant: CardVariant): string {
  const byRarity: Record<string, string> = {
    common: "hpkm-card--common",
    "uncommon reverse holo": "hpkm-card--reverse",
    "rare holo": "hpkm-card--holo",
    "rare holo cosmos": "hpkm-card--cosmos",
    "amazing rare": "hpkm-card--amazing",
    "radiant rare": "hpkm-card--radiant",
    "rare holo v": "hpkm-card--v",
    "rare ultra": "hpkm-card--vfa",
    "rare holo vmax": "hpkm-card--vmax",
    "rare rainbow": "hpkm-card--vmax-r",
    "rare holo vstar": "hpkm-card--vstar",
    "rare secret": "hpkm-card--gold",
  };
  return byRarity[variant.dataRarity] ?? "hpkm-card--common";
}

export function wireSubtypes(variant: CardVariant, stage: string): string {
  if (variant.trainerGallery || variant.subtypes === "radiant") {
    return variant.subtypes;
  }
  if (stage === "Stage 1") return "stage1";
  if (stage === "Stage 2") return "stage2";
  return variant.subtypes || "supporter";
}

export function wireNumber(variant: CardVariant, level: number): string {
  if (variant.trainerGallery) return `TG${String(level).padStart(2, "0")}`;
  return String(level);
}

/** TCG outer frame tier — yellow basics, black V, grey full-art, none for VMAX */
export type CardBorderTier = "basic-yellow" | "v-black" | "full-art-grey" | "none" | "secret-gold";

export function cardBorderTier(variant: CardVariant): CardBorderTier {
  switch (variant.faceTemplate) {
    case "trainer":
      return "basic-yellow";
    case "pokemon-v":
      return "v-black";
    case "vmax":
    case "rainbow":
      return "none";
    case "secret":
      return "secret-gold";
    case "full-art":
    case "vstar":
    default:
      return "full-art-grey";
  }
}

/** CSS root class for holo clipping behaviour */
export function faceLayoutClass(variant: CardVariant): string {
  return variant.faceTemplate === "trainer" ? "hpk-trainer-face" : "hpk-full-bleed";
}
